src/transformers/quantizers/quantizer_bnb_4bit.py

```diff
@@ -142,6 +142,11 @@ def check_quantized_param(
         module, tensor_name = get_module_from_name(model, param_name)
         if isinstance(module._parameters.get(tensor_name, None), bnb.nn.Params4bit):
             # Add here check for loaded components' dtypes once serialization is implemented
+            if self.pre_quantized:
+                if param_value.dtype != torch.uint8:
+                    raise ValueError(
+                        f"Incompatible dtype `{param_value.dtype}` when loading 4-bit prequantized weight. Expected `torch.uint8`."
+                    )
```

Comment:
In which case this happens ? Is there a related issue ? It would be nice to propose a workaround to the user, for example quantize the model on the fly instead because this one doesn't work for some reason

Reply:
I guess it happens if there is some issue with the quantized model during quantization, it's the same check in bnb 8-bit. I can add a more meaningful message for both quantizers

Reply:
In particular for use with QLoRA+FSDP, it's valid to store in a dtype other than uint8. The expected dtype would be part of the quantization config.

Some documentation this: https://huggingface.co/docs/bitsandbytes/en/fsdp_qlora

Example model serialized this way: https://huggingface.co/hugging-quants/Meta-Llama-3.1-70B-BNB-NF4-BF16

Reply:
Thanks @matthewdouglas ! In the quantization config of the example model :
```
"quantization_config": {
    "_load_in_4bit": true,
    "_load_in_8bit": false,
    "bnb_4bit_compute_dtype": "bfloat16",
    "bnb_4bit_quant_storage": "bfloat16",
    "bnb_4bit_quant_type": "nf4",
    "bnb_4bit_use_double_quant": true,
    "llm_int8_enable_fp32_cpu_offload": false,
    "llm_int8_has_fp16_weight": false,
    "llm_int8_skip_modules": null,
    "llm_int8_threshold": 6.0,
    "load_in_4bit": true,
    "load_in_8bit": false,
    "quant_method": "bitsandbytes"
  }
```
I can't find the dtype here

Reply:
@MekkCyber The `"bnb_4bit_quant_storage": "bfloat16"` indicates that a bfloat16 tensor is used as the underlying storage.

Reply:
Ah okay thanks @matthewdouglas ! And does it make sense to add this check with `bnb_4bit_quant_storage` instead of just hardcoding `uint8` ?

Reply:
@MekkCyber It's not important enough to be worth the effort IMO.